### PR TITLE
Database Layer complete

### PR DESCRIPTION
**Pull request by Hammad887**
Used MySQL database.
Includes CRUD operations

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+	@echo
+	@echo 'Commands:'
+	@echo '  build           		Compile project.'
+	@echo '  check           		Run linters.'
+	@echo '  format          		Format source code.'
+	@echo '  codegen         		Generate code.'
+	@echo '  test            		Run test case'
+	@echo '  prepare         		build dev container'
+	@echo '  clean       			remove dev temp folder'
+	@echo
```
Comment (shahzadhaider1) [inline Makefile:47]: please leave blank line at the end

File: Makefile.variables [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+# all non-windows environments
+ROOT := $(shell pwd)
+
+DEV_IMAGE := ${PROJECT}_dev
+
+DOCKERRUN := docker run --rm \
+	-v ${ROOT}:/usr/src/${PROJECT} \
+	-w /usr/src/${PROJECT} \
+	${DEV_IMAGE}
```
Comment (shahzadhaider1) [inline Makefile.variables:14]: please leave blank line at the end

File: Makefile.variables [hunk 3]
```diff
@@ -0,0 +1,14 @@
+export PROJECT = chat-app
+IMPORT_PATH = github.com/Hammad887/${PROJECT}
+
+MYSQL_DB_CNT_ID ?=
```
Comment (shahzadhaider1) [inline Makefile.variables:4]: what's the purpose of this?

File: config/config.go [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+import (
+	"fmt"
+
+	"github.com/spf13/viper"
+)
+
+// keys for database configuration
+const (
+	MySQLDBHost     = "MYSQL_DB_HOSTS"
```
Comment (shahzadhaider1) [inline config/config.go:11]: HOSTS or HOST?

File: config/config.go [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+	viper.SetConfigType("env")
+	viper.SetConfigFile(".env") // Assuming you're using .env in the root
+	viper.AutomaticEnv()
+
+	err := viper.ReadInConfig()
+	fmt.Println(viper.AllSettings())
+
+	if err != nil {
+		fmt.Printf("Error reading config file: %s\n", err)
+	}
```
Comment (shahzadhaider1) [inline config/config.go:27]: we always leave a blank line after error handling

File: db/mysql/client.go [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+	ctx := context.Background()
+
+	username := viper.GetString(config.MySQLDBUsername)
+	password := viper.GetString(config.MySQLDBPassword)
+	hostname := viper.GetString(config.MySQLDBHost)
+	dbName := viper.GetString(config.MySQLDBName)
+	// username := "hammad"
+	// password := "Hammad_887"
+	// hostname := "127.0.0.1:3306"
+	// dbName := "chatapp"
```
Comment (shahzadhaider1) [inline db/mysql/client.go:31]: remove these comments

File: db/mysql/client.go [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+}
+
+// NewClient creates a configured database client.
+func NewClient(option db.Option) (db.DataStore, error) {
+	ctx := context.Background()
+
+	username := viper.GetString(config.MySQLDBUsername)
+	password := viper.GetString(config.MySQLDBPassword)
+	hostname := viper.GetString(config.MySQLDBHost)
+	dbName := viper.GetString(config.MySQLDBName)
```
Comment (shahzadhaider1) [inline db/mysql/client.go:27]: please format your code

File: db/mysql/client.go [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+	// username := "hammad"
+	// password := "Hammad_887"
+	// hostname := "127.0.0.1:3306"
+	// dbName := "chatapp"
+
+	fmt.Println(dsn(username, password, hostname, dbName))
```
Comment (shahzadhaider1) [inline db/mysql/client.go:33]: we shouldn't print our credentials

File: db/mysql/client.go [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+
+	// dbClient, err := sql.Open("mysql", "abc:password@tcp(db:3306)/test_database")
+
+	dbClient, err := sql.Open("mysql", "hammad:Hammad_887@tcp(localhost:3306)/chatapp")
```
Comment (shahzadhaider1) [inline db/mysql/client.go:37]: we shouldn’t hard-code the credentials

File: db/mysql/client.go [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+
+	// dbClient, err := sql.Open("mysql", "newuser:password@tcp(db:3306)/chatapp")
```
Comment (shahzadhaider1) [inline db/mysql/client.go:39]: this comment is also irrelevant

File: db/mysql/client.go [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+	}
+
+	if err := dbClient.PingContext(ctx); err != nil {
+		return nil, wraperrors.Wrap(err, "unable to ping database")
+	}
+
+	return &client{dbc: dbClient}, nil
+}
+
+func dsn(username string, password string, hostname string, dbName string) string {
```
Comment (shahzadhaider1) [inline db/mysql/client.go:51]: do we need this function?

File: db/mysql/ops.go [hunk 12]
```diff
@@ -0,0 +1,264 @@
...
+	"database/sql"
+	"errors"
+	"fmt"
+	"log"
+	"net/http"
+	"os"
+	"strings"
+	"time"
+
+	domain "github.com/Hammad887/chat-app/models"
```
Comment (shahzadhaider1) [inline db/mysql/ops.go:14]: please sort imports this way:

1. local packages
2. third party packages 
3. project packages

File: db/mysql/ops.go [hunk 13]
```diff
@@ -0,0 +1,264 @@
...
+	hashedPassword, err := bcrypt.GenerateFromPassword([]byte(user.Password), bcrypt.DefaultCost)
+	if err != nil {
+		return false, err
+	}
+
+	user.ID = uuid.New().String()
+
+	if _, err := c.dbc.Exec("INSERT INTO users (id, name, email, password) VALUES (?, ?, ?, ?)", user.ID, user.Name, user.Email, string(hashedPassword)); err != nil {
+		return false, err
+
```
Comment (shahzadhaider1) [inline db/mysql/ops.go:31]: we can remove this blank line

File: db/mysql/ops.go [hunk 14]
```diff
@@ -0,0 +1,264 @@
...
+	message.ID = uuid.New().String()
+
+	// c.assignUserToChatRoom()
+
+	_, err := c.dbc.Exec("INSERT INTO messages (id, text, sender_id, room_id, created_at) VALUES (?, ?, ?, ?, ?)", message.ID, message.Text, message.SenderID, message.RoomID, time.Now())
+	if err != nil {
+		log.Println(err)
+		return false, err
+	}
```
Comment (shahzadhaider1) [inline db/mysql/ops.go:48]: blank line after error handling

File: db/mysql/ops.go [hunk 15]
```diff
@@ -0,0 +1,264 @@
...
+
+	tokenString, err := token.SignedString(jwtKey)
+	if err != nil {
+		return "", err
+	}
+
+	return tokenString, nil
+}
+
+func tokenVerificationMiddleware(next http.Handler) http.Handler {
```
Comment (shahzadhaider1) [inline db/mysql/ops.go:204]: why is this handler inside the database layer?

File: errors/error.go [hunk 16]
```diff
@@ -0,0 +1,10 @@
+package errors
+
+import "errors"
+
+var (
+	// ErrNoContent custom error for no record found
+	ErrNoContent = errors.New("no model exist in database")
+	// ErrConflict custom error for record already exist
+	ErrConflict = errors.New("user model is already exist in database")
```
Comment (shahzadhaider1) [inline errors/error.go:9]: you have models other than user as well. this should be generic

File: models/chatroom.go [hunk 17]
```diff
@@ -0,0 +1,7 @@
+package models
+
+type ChatRoom struct {
```
Comment (shahzadhaider1) [inline models/chatroom.go:3]: please add comments for exported structs

**Review by shahzadhaider1 — CHANGES_REQUESTED** [review]

File: config/config.go (hunk 5 above)
Comment (Hammad887) [reply config/config.go:27]: Got it. Changed!

File: models/chatroom.go (hunk 17 above)
Comment (Hammad887) [reply models/chatroom.go:3]: Added.

File: Makefile (hunk 1 above)
Comment (Hammad887) [reply Makefile:47]: Done

File: Makefile.variables (hunk 2 above)
Comment (Hammad887) [reply Makefile.variables:14]: Done

File: Makefile.variables (hunk 3 above)
Comment (Hammad887) [reply Makefile.variables:4]: This was to set up the environment variables and paths.
This was extra. I have removed it.

File: config/config.go (hunk 4 above)
Comment (Hammad887) [reply config/config.go:11]: It should be HOST. Changed it

File: db/mysql/client.go (hunk 6 above)
Comment (Hammad887) [reply db/mysql/client.go:31]: Removed

File: db/mysql/client.go (hunk 7 above)
Comment (Hammad887) [reply db/mysql/client.go:27]: Done. Used gofmt -w . and goimports -w .

File: db/mysql/client.go (hunk 8 above)
Comment (Hammad887) [reply db/mysql/client.go:33]: You are right. This was just for testing. Removed it

File: db/mysql/client.go (hunk 9 above)
Comment (Hammad887) [reply db/mysql/client.go:37]: Changed them to environment variables

File: db/mysql/client.go (hunk 10 above)
Comment (Hammad887) [reply db/mysql/client.go:39]: Deleted

File: db/mysql/client.go (hunk 11 above)
Comment (Hammad887) [reply db/mysql/client.go:51]: It isn't necessary, but I am using it in the NewClient function. So it's just acting as a helper function.

File: db/mysql/ops.go (hunk 12 above)
Comment (Hammad887) [reply db/mysql/ops.go:14]: Sorted

File: db/mysql/ops.go (hunk 13 above)
Comment (Hammad887) [reply db/mysql/ops.go:31]: Removed

File: db/mysql/ops.go (hunk 14 above)
Comment (Hammad887) [reply db/mysql/ops.go:48]: Added blank lines wherever I am handling errors now.

File: db/mysql/ops.go (hunk 15 above)
Comment (Hammad887) [reply db/mysql/ops.go:204]: Left here by mistake. Removed now

File: errors/error.go (hunk 16 above)
Comment (Hammad887) [reply errors/error.go:9]: Done

**Comment by Hammad887** [comment]
Done all the changes. Thank you for the reviews!

**Pull request merged**